kvserver: add rac2 pull mode integration test variations

kvoli opened the pull request:
First commit from https://github.com/cockroachdb/cockroach/pull/132121
Second commit from https://github.com/cockroachdb/cockroach/pull/132125

---

Introduce `pull_mode=true` variants to the existing
`TestFlowControl.*V2` integration tests.

These tests use their own file, as the behavior differs in some cases
between pull and push mode.

Part of: https://github.com/cockroachdb/cockroach/issues/128040
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132209)
<!-- Reviewable:end -->

kvoli commented:
I've mixed in some changes in the last commit that belong in the first commit -- the first commit was failing tests because:

- The v1 flow control integration tests were now using the v2 machinery due to not overriding the cluster version


The second commit was failing tests because:

- pull mode could be enabled and the v2 (and v1) flow control integration tests were issuing normal priority writes while also blocking the store work queue from granting admission -- hanging the test

There didn't appear to be any non-test code changes required to get this all to work. As for whether the results look reasonable, that remains to be seen.

I will clean up the commits shortly.

kvoli committed:
kvserver: allow enabling rac pull mode with a send queue

By default,`kvadmission.flow_control.mode = "apply_to_elastic"`
replication admission control operates in push mode, where Raft has its
own flow control machinery in operation and controls sending messages to
replicas.

In pull mode, which could not be enabled prior to this commit (outside
of tests), replication flow control (v2) maintains its own machinery
(send queue) and controls when to send a message to a replica.

Previously, when `kvadmission.flow_control.mode = "apply_to_all"`,
replication admission control would remain in push mode.

Allow replication flow control to run in pull mode outside of tests, by
setting both:

```
kvadmission.flow_control.enabled = true (default true)
kvadmission.flow_control.mode = "apply_to_all" (default "apply_to_elastic")
```

And running a cluster which has finalized its upgrade to `v24.3`.

We omit a release note, as this functionality is disabled by default and
we wish to retain the ability to selectively suggest enabling pull mode
to users, as the feature is (1) newly introduced, (2) on the critical
path for all replication and (3) subject to change.

Resolves: #132115
Release note: None

kvoli committed:
kvserver: specify flow_control_integration_test_v2 encoding

Note to the reader, only files have moved without other changes.

We didn't specify the v2 encoding in the testdata file but did for v1
encoding.

Now specify both.

Part of: #128040
Release note: None

kvoli committed:
kvserver: add rac2 pull mode integration test variations

Introduce pull_mode=true variants to the existing
TestFlowControl.*V2 integration tests.

These tests use their own file, as the behavior differs in some cases
between pull and push mode.

Part of: #128040
Release note: None

kvoli committed:
kvserver: initialize store config kvflow fields if test

Some tests don't setup a server, just a store or similar. In these
cases, the KVFlow.* fields in the store config, which are references
to a shared object between stores on the same server, will not be set.
We set them here to avoid handling nil values in such tests within the
rac2 code.

Part of: #128040
Release note: None

kvoli committed:
kvserver: enable v2 enabled when leader level

`V2EnabledWhenLeaderLevel` captures the level at which RACv2 is enabled
when a replica is the leader. It previously was hardcoded to only return
`V2NotEnabledWhenLeader`, as the underlying implementation was still in
progress at the time.

The underlying implementation is now ready. Enable returning the correct
`V2EnabledWhenLeaderLevel` from `GetV2EnabledWhenLeaderLevel`.

Part of: #132115
Release note: None

kvoli commented:
Going to close this one in favor of https://github.com/cockroachdb/cockroach/pull/132125, since we need the tests before allowing pull mode to be enabled and those tests also require pull mode.
